File: Runtime/STSimMapBase5.vb
'************************************************************************************
' ST-Sim: A .NET class library for developing state-and-transition simulation models.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'************************************************************************************

Imports SyncroSim.Core
Imports SyncroSim.Common

MustInherit Class STSimMapBase5(Of T)
    Inherits STSimMapBase

    Private m_map As New MultiLevelKeyMap5(Of SortedKeyMap2(Of T))

    Protected Sub New(ByVal scenario As Scenario)
        MyBase.New(scenario)
    End Sub

    Protected Sub AddItem(
        ByVal k1 As Nullable(Of Integer),
        ByVal k2 As Nullable(Of Integer),
        ByVal k3 As Nullable(Of Integer),
        ByVal k4 As Nullable(Of Integer),
        ByVal k5 As Nullable(Of Integer),
        ByVal iteration As Nullable(Of Integer),
        ByVal timestep As Nullable(Of Integer),
        ByVal item As T)

        Dim m As SortedKeyMap2(Of T) = Me.m_map.GetItemExact(k1, k2, k3, k4, k5)

        If (m Is Nothing) Then

            m = New SortedKeyMap2(Of T)(SearchMode.ExactPrev)
            Me.m_map.AddItem(k1, k2, k3, k4, k5, m)

        End If

        Dim v As T = m.GetItemExact(iteration, timestep)

        If (v IsNot Nothing) Then
            ThrowDuplicateItemException()
        End If

        m.AddItem(iteration, timestep, item)
        Me.SetHasItems()

    End Sub

    Protected Function GetItemExact(
        ByVal k1 As Nullable(Of Integer),
        ByVal k2 As Nullable(Of Integer),
        ByVal k3 As Nullable(Of Integer),
        ByVal k4 As Nullable(Of Integer),
        ByVal k5 As Nullable(Of Integer),
        ByVal iteration As Nullable(Of Integer),
        ByVal timestep As Nullable(Of Integer)) As T

        Dim m As SortedKeyMap2(Of T) = Me.m_map.GetItemExact(k1, k2, k3, k4, k5)

        If (m Is Nothing) Then
            Return Nothing
        End If

        Return m.GetItemExact(iteration, timestep)

    End Function

    Protected Function GetItem(
        ByVal k1 As Nullable(Of Integer),
        ByVal k2 As Nullable(Of Integer),
        ByVal k3 As Nullable(Of Integer),
        ByVal k4 As Nullable(Of Integer),
        ByVal k5 As Nullable(Of Integer),
        ByVal iteration As Nullable(Of Integer),
        ByVal timestep As Nullable(Of Integer)) As T

        If (Not Me.HasItems) Then
            Return Nothing
        End If

        Dim p As SortedKeyMap2(Of T) = Me.m_map.GetItem(k1, k2, k3, k4, k5)

        If (p Is Nothing) Then
            Return Nothing
        End If

        Return p.GetItem(iteration, timestep)

    End Function

End Class

File: src/Runtime/STSimMapBase5.cs
// stsim: A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2023 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using SyncroSim.Core;
using SyncroSim.Apex;

namespace SyncroSim.STSim
{
    internal abstract class STSimMapBase5<T> : STSimMapBase
    {
        private MultiLevelKeyMap5<SortedKeyMap2<T>> m_map = new MultiLevelKeyMap5<SortedKeyMap2<T>>();

        protected STSimMapBase5(Scenario scenario) : base(scenario)
        {
        }

        protected void AddItem(int? k1, int? k2, int? k3, int? k4, int? k5, int? iteration, int? timestep, T item)
        {
            SortedKeyMap2<T> m = this.m_map.GetItemExact(k1, k2, k3, k4, k5);

            if (m == null)
            {
                m = new SortedKeyMap2<T>(SearchMode.ExactPrev);
                this.m_map.AddItem(k1, k2, k3, k4, k5, m);
            }

            T v = m.GetItemExact(iteration, timestep);

            if (v != null)
            {
                ThrowDuplicateItemException();
            }

            m.AddItem(iteration, timestep, item);
            this.SetHasItems();
        }

        protected T GetItemExact(int? k1, int? k2, int? k3, int? k4, int? k5, int? iteration, int? timestep)
        {
            SortedKeyMap2<T> m = this.m_map.GetItemExact(k1, k2, k3, k4, k5);

            if (m == null)
            {
                return default(T);
            }

            return m.GetItemExact(iteration, timestep);
        }

        protected T GetItem(int? k1, int? k2, int? k3, int? k4, int? k5, int? iteration, int? timestep)
        {
            if (!this.HasItems)
            {
                return default(T);
            }

            SortedKeyMap2<T> p = this.m_map.GetItem(k1, k2, k3, k4, k5);

            if (p == null)
            {
                return default(T);
            }

            return p.GetItem(iteration, timestep);
        }
    }
}
